### PR TITLE
Fixed Typehint for RunSettings.colocated_db_settings

### PR DESCRIPTION
**Pull request by amandarichardsonn**
This PR adds Python type hinting to RunSettings.colocated_db_settings

### CODE REVIEW

File: smartsim/entity/model.py [hunk 1]
```diff
@@ -328,13 +329,23 @@ def colocate_db_tcp(
...
         }
+        # test
         self._set_colocated_db_settings(tcp_options, common_options, **kwargs)
 
     def _set_colocated_db_settings(
         self,
-        connection_options: t.Dict[str, t.Any],
-        common_options: t.Dict[str, t.Any],
-        **kwargs: t.Any,
+        connection_options: t.Mapping[str, t.Union[int, list[str], str]],
```
Comment (MattToast) [inline smartsim/entity/model.py:337]: ditto "pedantic, would rather update `t.List[T]` -> `list[T]` all at once in a future ticket" comment here

File: smartsim/entity/model.py [hunk 2]
```diff
@@ -258,11 +258,12 @@ def colocate_db_uds(
                 f"Invalid name for unix socket: {unix_socket}. Must only "
                 "contain alphanumeric characters or . : _ - /"
             )
-
-        uds_options = {
+        assert isinstance(unix_socket, str)
+        assert isinstance(socket_permissions, int)
```
Comment (MattToast) [inline smartsim/entity/model.py:262]: Can we remove the `assert` stmts here? It looks like mypy should already be able to eventuate these types correctly

File: smartsim/entity/model.py [hunk 3]
```diff
@@ -357,21 +368,40 @@ def _set_colocated_db_settings(
...
-        common_options["custom_pinning"] = self._create_pinning_string(
-            common_options["custom_pinning"], common_options["cpus"]
+        x = t.cast(
+            t.Optional[t.Iterable[t.Union[int, t.Iterable[int]]]],
+            common_options.get("custom_pinning"),
         )
-
-        colo_db_config = {}
+        e = t.cast(int, common_options.get("cpus"))
+        common_options["custom_pinning"] = self._create_pinning_string(x, e)
```
Comment (MattToast) [inline smartsim/entity/model.py:376]: Can we change the names of these of these intermediary variables to something a bit more descriptive (maybe `custom_pinning_` and `cpus_`)?

Or maybe just remove them entirely?
```py
common_options["custom_pinning"] = self._create_pinning_string(
    t.cast(
        t.Optional[t.Iterable[t.Union[int, t.Iterable[int]]]], 
        common_options["custom_pinning"], common_options["cpus"]
    ),
    t.cast(int, common_options.get("cpus"))
)
```

**Review by MattToast — CHANGES_REQUESTED** [review]
A couple of super pedantic nit picks, but otherwise looks about ready to go! 🚀

File: smartsim/entity/model.py [hunk 4]
```diff
@@ -258,11 +258,12 @@ def colocate_db_uds(
...
             )
-
-        uds_options = {
+        assert isinstance(unix_socket, str)
+        assert isinstance(socket_permissions, int)
+        uds_options: t.Dict[str, t.Union[int, str]] = {
             "unix_socket": unix_socket,
             "socket_permissions": socket_permissions,
-            "port": 0,  # This is hardcoded to 0 as recommended by redis for UDS
+            "port": 0,  # 0 as recommended by redis for UDS
```
[the remark this answers is not in the document]
Comment (amandarichardsonn) [reply smartsim/entity/model.py:266]: I was trying to meet the line count, but I did it the lazy way and deleted words, I moved it to a new line

File: smartsim/entity/model.py [hunk 5]
```diff
@@ -328,13 +329,23 @@ def colocate_db_tcp(
             "debug": debug,
             "db_identifier": db_identifier,
         }
+        # test
```
[the remark this answers is not in the document]
Comment (amandarichardsonn) [reply smartsim/entity/model.py:332]: naur

File: smartsim/settings/lsfSettings.py [hunk 6]
```diff
@@ -97,7 +97,8 @@ def set_cpus_per_rs(self, cpus_per_rs: int) -> None:
         :type cpus_per_rs: int or str
         """
         if self.colocated_db_settings:
-            db_cpus = int(self.colocated_db_settings.get("db_cpus", 0))
+            db_cpus = self.colocated_db_settings.get("db_cpus", 0)
+            assert isinstance(db_cpus, int)
```
[the remark this answers is not in the document]
Comment (amandarichardsonn) [reply smartsim/settings/lsfSettings.py:101]: If Im already casting this to an int, why again?

File: smartsim/entity/model.py (hunk 3 above)
Comment (amandarichardsonn) [reply smartsim/entity/model.py:376]: I would almost suggest keeping the vars for readability, but Im not as experienced so let me know!

File: smartsim/settings/lsfSettings.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MattToast) [reply smartsim/settings/lsfSettings.py]: `t.cast` is purely for the type checker while the `int` will actually do the cast at runtime. It's HIGHLY unlikely, bu there is the possibility that the `"db_cpus"` key maps to a `float` (or a  `str`,  or some other type) that can be cast to an `int`.

so for instance
```py
self.colocated_db_settings = {"db_cpus": 3.0, ...}
db_cpus = int(self.colocated_db_settings.get("db_cpus", 0))
print(db_cpus)
```
will run just fine and print `3`

while
```py
self.colocated_db_settings = {"db_cpus": 3.0, ...}
db_cpus = self.colocated_db_settings.get("db_cpus", 0)
assert isinstance(db_cpus, int)
print(db_cpus)
```

will raise an `AssertionError`.


Just to make sure that this is preserving the original behavior and prevent an API break I would rather just subtly lie to the type checker and let it think that `self.colocated_db_settings.get("db_cpus", 0)` is some type that can be cast to an `int`. But do LMK if you think I am being too paranoid!

File: smartsim/entity/model.py (hunk 3 above)
Comment (MattToast) [reply smartsim/entity/model.py]: Totally willing to keep the intermediate variables! I think I agree that they improve readability!

File: smartsim/settings/lsfSettings.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (amandarichardsonn) [reply smartsim/settings/lsfSettings.py]: I see I see, some witchery I would say

**Pull request merged**